TYPO3.Flow/Classes/TYPO3/Flow/Utility/Lock/FlockLockStrategy.php

```diff
@@ -51,56 +55,97 @@ class FlockLockStrategy implements LockStrategyInterface
-            Files::createDirectoryRecursively($temporaryDirectory);
-            self::$temporaryDirectory = $temporaryDirectory;
+            $this->configureTemporaryDirectory();
         }
-        $this->lockFileName = Files::concatenatePaths(array(self::$temporaryDirectory, md5($subject)));
 
-        if (($this->filePointer = @fopen($this->lockFileName, 'r')) === false) {
-            if (($this->filePointer = @fopen($this->lockFileName, 'w')) === false) {
-                throw new LockNotAcquiredException(sprintf('Lock file "%s" could not be opened', $this->lockFileName), 1386520596);
-            }
+        $this->lockFileName = Files::concatenatePaths([self::$temporaryDirectory, md5($subject)]);
+        $continueTrying = true;
+        while ($continueTrying) {
+            $continueTrying = $this->tryToAquireLock($exclusiveLock);
+        }
+    }
+
+    /**
+     * Sets the temporaryDirectory as static variable for the lock class.
+     *
+     * @throws LockNotAcquiredException
+     * @throws \TYPO3\Flow\Utility\Exception
+     * return void;
+     */
+    protected function configureTemporaryDirectory()
+    {
+        if (Bootstrap::$staticObjectManager === null || !Bootstrap::$staticObjectManager->isRegistered(\TYPO3\Flow\Utility\Environment::class)) {
+            throw new LockNotAcquiredException('Environment object could not be accessed', 1386680952);
+        }
+        $environment = Bootstrap::$staticObjectManager->get(\TYPO3\Flow\Utility\Environment::class);
+        $temporaryDirectory = Files::concatenatePaths([$environment->getPathToTemporaryDirectory(), 'Lock']);
+        Files::createDirectoryRecursively($temporaryDirectory);
+        self::$temporaryDirectory = $temporaryDirectory;
+    }
+
+    /**
+     * Tries to open a lock file and apply the lock to it.
+     *
+     * @param boolean $exclusiveLock
+     * @return boolean True if a lock could not be aquired, false if the lock was successful.
+     * @throws LockNotAcquiredException
+     */
+    protected function tryToAquireLock($exclusiveLock)
+    {
+        if (($this->filePointer = @fopen($this->lockFileName, 'r')) === false && ($this->filePointer = @fopen($this->lockFileName, 'w')) === false) {
+            throw new LockNotAcquiredException(sprintf('Lock file "%s" could not be opened', $this->lockFileName), 1386520596);
         }
+        $this->applyFlock($exclusiveLock);
+        $fstat = fstat($this->filePointer);
+        $stat = stat($this->lockFileName);
+        // Make sure that the file did not get unlinked between the fopen and the actual flock
+        // This will always be TRUE on windows, because 'ino' stat will always be 0, but unlink is not possible on opened files anyway
+        if ($stat !== false && $stat['ino'] === $fstat['ino']) {
+            return false;
+        }
+
+        flock($this->filePointer, LOCK_UN);
+        fclose($this->filePointer);
+        $this->filePointer = null;
+
+        usleep(100 + rand(0, 100));
```

Comment:
This usleep can be moved up before the return false, since it's only needed if locking failed and will be reattempted, in order to prevent deadlocks due to two processes synchronously retrying to acquire the same lock.